### PR TITLE
Protect against cache corruption

### PR DESCRIPTION
**Pull request by zdrve**
I still occasionally see cache corruption:

```
$ with-ruby - bundle exec rake generate kennel:update_datadog PROJECT=guide-redirect-service
Generating ... 2.05s
Storing ... 0.0s
Downloading definitions ... 6.6s
Diffing ... -rake aborted!
NoMethodError: undefined method `reject!' for nil:NilClass
/Users/revans/git/github.com/grosser/kennel/lib/kennel/file_cache.rb:60:in `expire_old_data'
/Users/revans/git/github.com/grosser/kennel/lib/kennel/file_cache.rb:19:in `open'
...

$ mv tmp/cache tmp/cache.bad
$ with-ruby - bundle exec rake generate kennel:update_datadog PROJECT=guide-redirect-service
<works fine>
```

In this case, somehow a `nil` had been persisted. Not sure how, but here's a defence: 23e3842

Another form of possible corruption is partial writes. Fix: 75ff2dec0eed79f492cf8394e2f7c4ad05350e6c

### CODE REVIEW

File: lib/kennel/file_cache.rb [hunk 1]
```diff
@@ -44,11 +44,14 @@ def load_data
...
 
     def persist
+      return if @data.nil?
+
       dir = File.dirname(@file)
       FileUtils.mkdir_p(dir) unless File.directory?(dir)
 
       Tempfile.create "kennel-file-cache", dir do |tmp|
         Marshal.dump @data, tmp
+        tmp.flush
```
Comment (grosser) [inline lib/kennel/file_cache.rb:54]: that's the good part I think 🤞

Comment (grosser) [reply lib/kennel/file_cache.rb:54]: ... actually `.close` would be better

File: lib/kennel/file_cache.rb [hunk 2]
```diff
@@ -44,11 +44,14 @@ def load_data
     end
 
     def persist
+      return if @data.nil?
```
Comment (grosser) [inline lib/kennel/file_cache.rb:47]: prefer to make load_data fallback to `{}` we should never have nil data here

File: lib/kennel/file_cache.rb (hunk 1 above)
Comment (zdrve) [reply lib/kennel/file_cache.rb:54]: No difference really, it's going to close anyway

File: lib/kennel/file_cache.rb (hunk 2 above)
Comment (zdrve) [reply lib/kennel/file_cache.rb:47]: OK, do have a specific patch to propose?

Comment (zdrve) [reply lib/kennel/file_cache.rb:47]: I have a _suspicion_ as to how my cache ended up containing simply a `nil` (the cache file was literally 3 bytes long, the exact same 3 bytes that `Marshal.dump(nil)` gives), i.e. in that specific case, it wasn't the non-flushing problem, it was the nil problem. However ... we kinda _already_ fall back to `{}`, right? Except it doesn't seem to work.

Comment (grosser) [reply lib/kennel/file_cache.rb:47]: on line 40 we should say `|| {}`

Comment (zdrve) [reply lib/kennel/file_cache.rb:47]: That's kinda missing the point, isn't it? That says: if the cache is `nil` (i.e. already corrupted), then try to un-corrupt it. Shouldn't we be trying to make sure it doesn't corrupt in the first place?

**Comment by grosser** [comment]
Yeah raise when it's about to store a nil so we can find the root cause 👍

**Comment by zdrve** [comment]
How would that help find the root cause? 

This is an interesting discussion but I'm not finding it very "actionable" when it comes to this PR.

File: lib/kennel/file_cache.rb [hunk 3]
```diff
@@ -15,11 +15,13 @@ def initialize(file, cache_version)
     end
 
     def open
-      load_data
-      expire_old_data
-      yield self
-    ensure
-      persist
```
Comment (grosser) [inline lib/kennel/file_cache.rb:22]: I bet that's it 🎉

File: lib/kennel/file_cache.rb (hunk 1 above)
Comment (grosser) [reply lib/kennel/file_cache.rb:51]: close is more final than flush, I doubt it makes a difference, but we want to say "ok we are done with this file"

**Comment by grosser** [comment]
yeah I bet the ensure block was the issue ... it crashes and then writes nil 🎉

**Pull request merged**